ZcashFoundation/zebra
# feat(net): Cache a list of useful peers on disk

## Motivation

Zebra currently uses DNS seeders as its only source of peers. To make it more secure and reliable, we store a list of recent useful peers on disk, and read it at startup.

Close #1956

### Specifications

Peer behaviour is not specified, but `zcashd` does something like this. (In fact it tries to rely on DNS even less.)

### Complex Code or Requirements

This PR introduces a new async task, and new filesystem code that must be run in an async context.

## Solution

- Add a new `zebra-network` config for the peer cache directory
- Regularly write recent useful peers to a file
    - Do it atomically so the file isn't corrupted if Zebra shuts down or panics while it is being written
    - If the peer cache would become empty, keep the previous cache file
    - Limit the number of peers in the peer cache
    - Add a new `zebra-network` peer set task to run this code 
- At startup, load the peers from a file, and use them in addition to DNS
- Add metrics and logging for the peer cache

Related changes:
- Update state cache docs to fix outdated info
- Add some zebra-chain and zebra-network convenience functions
- Panic if the configured peer addresses in `zebrad.toml` are invalid, rather than continuing

### Testing

Config:
- Add 3 new zebrad.toml config test files containing the variants of the new config field

Unit:
- make sure the peer cache is written by zebra_network::initialize and can be read back manually
- make sure the peer cache is written by zebra_network::initialize and is automatically loaded by the next instance

Acceptance:
- make sure the peer cache file path is written by `zebrad` in the existing persistent cache test
- make sure that the peer cache is not written when the cache is disabled (existing tests)
- make sure that a missing cache_dir config gets the default path, but a disabled config stays disabled when written to disk (existing tests)

Test Fixes:
- Update some existing tests to disable or configure the peer cache as needed
- Update the config test failure message for the new toml format using standard quotes

## Review

This is a routine audit fix.

Design questions:
- should cached peers replace DNS peers, or be used in combination with DNS peers?
- the peer cache is shared by all Zebra instances run by a single user, is this ok?

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

We might also want to add hard-coded IP addresses to the default list of peers.
